cpp/src/arrow/json/parser.cc

```diff
@@ -383,13 +392,14 @@ class RawArrayBuilder<Kind::kObject> {
 
  private:
   struct FieldInfo {
-    string_view name;  // Backed by key's allocation in name_to_index_
+    std::unique_ptr<std::string> name;
     BuilderPtr builder;
   };
```

Comment:
The simplest solution I could think of to avoid the `std::string` relocation issues. The faster approach would be to store the allocations in a separate `std::forward_list` and hold another `string_view` here, avoiding the double-dereference (at a higher memory cost).

Reply:
A separate list sounds better. It might not even be much more costly (in both cases you have one extra allocation per string).

Reply:
I'm inclined to agree - seems like the right tradeoff.